src/NuGet.Core/NuGet.Configuration/Settings/SettingFactory.cs

```diff
@@ -96,13 +97,84 @@ internal static SettingBase Parse(XNode node, SettingsFile origin)
             return null;
         }
 
+        private class SettingElementKeyComparer : IComparer<SettingElement>, IEqualityComparer<SettingElement>
+        {
+            public int Compare(SettingElement x, SettingElement y)
+            {
+                if (ReferenceEquals(x, y))
+                {
+                    return 0;
+                }
+
+                if (ReferenceEquals(x, null))
+                {
+                    return -1;
+                }
+
+                if (ReferenceEquals(y, null))
+                {
+                    return 1;
+                }
+
+                return StringComparer.OrdinalIgnoreCase.Compare(
+                    x.ElementName + string.Join("", x.Attributes.Select(a => a.Value)),
+                    y.ElementName + string.Join("", y.Attributes.Select(a => a.Value)));
+            }
+
+            public bool Equals(SettingElement x, SettingElement y)
+            {
+                if (ReferenceEquals(x, y))
+                {
+                    return true;
+                }
+
+                if (ReferenceEquals(x, null) || ReferenceEquals(y, null))
+                {
+                    return false;
+                }
+
+                return StringComparer.OrdinalIgnoreCase.Equals(
+                    x.ElementName + string.Join("", x.Attributes.Select(a => a.Value)),
+                    y.ElementName + string.Join("", y.Attributes.Select(a => a.Value)));
+            }
+
+            public int GetHashCode(SettingElement obj)
+            {
+                if (ReferenceEquals(obj, null))
+                {
+                    return 0;
+                }
+
+                return StringComparer.OrdinalIgnoreCase.GetHashCode(obj.ElementName + string.Join("", obj.Attributes.Select(a => a.Value)));
+            }
+        }
+
         internal static IEnumerable<T> ParseChildren<T>(XElement xElement, SettingsFile origin, bool canBeCleared) where T : SettingElement
         {
             var children = new List<T>();
+            IEnumerable<T> descendants = xElement.Elements().Select(d => Parse(d, origin)).OfType<T>();
+            SettingElementKeyComparer comparer = new SettingElementKeyComparer();
 
-            var descendants = xElement.Elements().Select(d => Parse(d, origin)).OfType<T>().Distinct();
+            HashSet<T> distinctDescendants = new HashSet<T>(comparer);
+
+            List<T> duplicatedDescendants = new List<T>();
```

Comment:
nit: Easy win to create the list only when needed since settings reading is actually part of VS init.